File: window.js
/**
 * This is the "find how many permutations of a little string are inside of a big string" problem
 * Utilies the "window" solution
 */

 const findPermutations = (string) => {
    const results = [];

    if (string.length === 1) 
    {
      results.push(string);
      return results;
    }

    for (var i = 0; i < string.length; i++) 
    {
      const firstChar = string[i];
      const otherChar = string.substring(0, i) + string.substring(i + 1);
      const otherPermutations = findPermutations(otherChar);

      for (var j = 0; j < otherPermutations.length; j++) {
        results.push(firstChar + otherPermutations[j]);
      }
    }
    return results;
 }
/**To fix by eod: not catching last permutation because it is short circuiting when end === bigString.length */
 const stringWithinString = (bigString, smallString) => {
     const hash = {};
     const permutations = findPermutations(smallString);
     for (let i = 0; i < permutations.length; i++) {
         hash[permutations[i]] = true
     }
     let start = 0;
     let end = smallString.length;
     let count = 0;
     let window;
     while (end < bigString.length) {
        window = bigString.slice(start, end);
        if (hash[window]) {
            count += 1
        } 
        start += 1; 
        end += 1;
     }
     return count;
 }

 console.log(stringWithinString("buttersnugglebuttkooooooobutt", "butt")) 

File: window.rb
 def permutations (string)
    results = [];

    if (string.length == 1) 
      results << string
      return results;
    end

    (0...string.length).each do |i| 
        first_char = string[i];
        other_char = string[0...i] + string[(i + 1)..-1];
        other_permutations = permutations(other_char);

        (0...other_permutations.length).each do |j| 
            results << (first_char + other_permutations[j]);
        end
    end
    results
end

def string_within_string(big, small)
    return 0 if big.length < small.length
    return 1 if big == small
    hash = {};
    permutations = permutations(small);
    permutations.each {|perm| hash[perm] = true }
    start = 0;
    count = 0;
    finish = small.length;
    while (finish < big.length)
        window = big[start...finish];
        count += 1 if hash[window]
        puts window
        start += 1
        finish += 1
    end
    count
end

puts string_within_string("buttuber", "butt") #should be 2